### PR TITLE
Get Pose Tracking demo working

### PR DESCRIPTION
**Pull request by AdamPettinger**
### Description

Initial commit here just changes the `robot_link_command_frame` in the Servo parameters to be the `ee_link` so that pose tracking uses that for the error calculations.

Some things I am looking at knocking out:

- [x] 1. Somewhere in the params, name a EE link (which is != to `robot_link_command_frame` necessarily)
- [x] 2. Add functionality to Servo to return the pose of the EE link w.r.t. the planning frame (also named in params). Think this will be useful for people who use pose_tracking so they: get pose, modify it, and move to new one
- [x] 3. Update pose_tracking PID error calc to use the difference between EE frame (from 1.) and target, instead of `robot_link_command_frame` to target
- [x] 4. Update the pose_tracking demo to use smaller pose-differences, and start from the current pose
- [x] 5. Add dynamic pose tracking?
- [x] 6. Gotta have them sweet sweet GIFs

Pose tracking to a (small-delta) static pose:
static_target_gif

Pose tracking with a dynamic target pose:
dynamic_target_gif

### CODE REVIEW

**Comment by AdamPettinger** [comment]
@AndyZe I think I am done with this PR. I want to make some changes to PoseTracking, but I think this PR is of a good scope now. I put some gifs of the demo in the description above, but there doesn't exist instructions on how to run it anywhere (you need to plan-execute away from singular, do all the gazebo stuff, and run the demo executable)

The second gif is the current state of the demo

File: moveit_ros/moveit_servo/src/servo_calcs.cpp [hunk 1]
```diff
@@ -174,6 +174,15 @@ void ServoCalcs::start()
 {
   stop_requested_ = false;
   timer_ = nh_.createTimer(period_, &ServoCalcs::run, this);
+
+  // TODO(adamp or andyz): after rebasing on upstream master, update this section. Maybe put the tf updates in a tiny
+  // function by themselves
+  sensor_msgs::JointStateConstPtr latest_joint_state = joint_state_subscriber_->getLatest();
```
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:180]: Basically, this ensures these things are initialized?

I would be OK with deleting the comment above. This seems fine

File: moveit_ros/moveit_servo/src/cpp_interface_example/pose_tracking_example.cpp [hunk 2]
```diff
@@ -109,27 +112,44 @@ int main(int argc, char** argv)
...
   geometry_msgs::PoseStamped target_pose;
-  target_pose.header.frame_id = "base_link";
+  target_pose.header.frame_id = current_ee_tf.header.frame_id;
+  target_pose.pose.position.x = current_ee_tf.transform.translation.x;
+  target_pose.pose.position.y = current_ee_tf.transform.translation.y;
+  target_pose.pose.position.z = current_ee_tf.transform.translation.z;
+  target_pose.pose.orientation.x = current_ee_tf.transform.rotation.x;
+  target_pose.pose.orientation.y = current_ee_tf.transform.rotation.y;
+  target_pose.pose.orientation.z = current_ee_tf.transform.rotation.z;
+  target_pose.pose.orientation.w = current_ee_tf.transform.rotation.w;
```
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/cpp_interface_example/pose_tracking_example.cpp:128]: You could save a few lines by doing:

`target_pose.pose.orientation = current_ee_tf.transform.orientation`

It won't work for translation, though (one is a Point, the other is a Vector3)

File: moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h [hunk 3]
```diff
@@ -78,6 +79,17 @@ class ServoCalcs
...
+
+  /**
+   * Get the End Effector link transform.
+   * The transform from the MoveIt planning frame to EE link
+   *
+   * @param transform the transform that will be calculated
+   * @return true if a valid transform was available
+   */
+  bool getEEFrameTransform(Eigen::Isometry3d& transform);
+  bool getEEFrameTransform(geometry_msgs::TransformStamped& transform);
```
Comment (AndyZe) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:92]: if you wrote a function like:

`geometry_msgs::TransformStamped convertEigenIsometryToTransStamped(const Eigen::Isometry3d& eigen_tf, const std::string& parent_frame, const std::string& child_frame)`

Then the definition of `bool getEEFrameTransform(geometry_msgs::TransformStamped& transform);` would be really simple:

```
bool getEEFrameTransform(geometry_msgs::TransformStamped& transform)
{
  Eigen::Isometry3d eigen_tf;
  getEEFrameTransform(eigen_tf);
  transform = convertEigenIsometryToTransStamped(eigen_tf);
}
```

Plus you could use it for `bool getCommandFrameTransform(geometry_msgs::TransformStamped& transform);` too

File: moveit_ros/moveit_servo/src/servo_calcs.cpp [hunk 4]
```diff
@@ -981,6 +1001,85 @@ bool ServoCalcs::getCommandFrameTransform(Eigen::Isometry3d& transform)
...
+  // Use Eigen to convert rotation to quaternion and fill in
+  Eigen::Quaterniond quat(rotation);
+  transform.transform.rotation.x = quat.x();
+  transform.transform.rotation.y = quat.y();
+  transform.transform.rotation.z = quat.z();
+  transform.transform.rotation.w = quat.w();
+
+  // Fill in rest of TF info
+  transform.header.frame_id = parameters_.planning_frame;
+  transform.child_frame_id = parameters_.ee_frame_name;
```
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:1078]: I think there's a little bit of potential to reduce duplicate lines with `getCommandFrameTransform(geometry_msgs::TransformStamped& transform)`  (see comment above).

But I'm cool keeping it as-is, too

**Review by AndyZe — COMMENTED** [review]
Generally looks good, I'll test it now before approving it

**Comment by AndyZe** [comment]
Gave it a test run, looks good!

File: moveit_ros/moveit_servo/src/servo_calcs.cpp (hunk 1 above)
Comment (AdamPettinger) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:180]: After one of my recent PR's, a lot more stuff got added to the start function, including a call to `joint_state_subscriber_->getLatest()` which we should at least reuse.

I thought about rebasing onto master in this PR, but decided against it. There are some conflicts

File: moveit_ros/moveit_servo/src/cpp_interface_example/pose_tracking_example.cpp (hunk 2 above)
Comment (AdamPettinger) [reply moveit_ros/moveit_servo/src/cpp_interface_example/pose_tracking_example.cpp:128]: For sure! Fixed it

File: moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h (hunk 3 above)
Comment (AdamPettinger) [reply moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:92]: Yep, should have done that the first time, fixed!

**Comment by AdamPettinger** [comment]
I liked your recommendations, and updated accordingly!

**Pull request merged**